module/spock/spock_verifier.go

```diff
@@ -0,0 +1,191 @@
+		verified, err := crypto.SPOCKVerify(approver.StakingPubKey, approval.Body.Spock, executor.StakingPubKey, receipt.Spocks[approval.Body.ChunkIndex])
+		if err != nil {
+			return nil, fmt.Errorf("could not verify spocks: %w", err)
+		}
+		if verified {
+			// map doesnt exist create it
+			if _, ok = v.approvals[approval.Body.ExecutionResultID]; !ok {
+				v.approvals[approval.Body.ExecutionResultID] = make(map[flow.Identifier]*flow.ExecutionReceipt)
+			}
+
+			// cache result for future verify calls for this approval
+			v.approvals[approval.Body.ExecutionResultID][approval.ID()] = receipt
+
+			return receipt, nil
+		}
+	}
+
+	return nil, nil
+}
+
+// ClearReceipts clears all receipts for a specific resultID
+func (v *Verifier) ClearReceipts(resultID flow.Identifier) bool {
+	v.Lock()
+	defer v.Unlock()
+
+	// check if entry exists
+	_, ok := v.receipts[resultID]
+	if !ok {
+		return false
+	}
+	_, ok = v.approvals[resultID]
+	if !ok {
+		return false
+	}
+
+	// clear receipts and approval cache
+	delete(v.receipts, resultID)
+	delete(v.approvals, resultID)
+
+	return true
+}
+
+func (v *Verifier) matchReceipt(receipt *flow.ExecutionReceipt) (bool, error) {
+	unmatchedReceipts := v.receipts[receipt.ExecutionResult.ID()]
+
+	// get idenitity of candidate receipt
+	identity, err := v.protocolState.AtBlockID(receipt.ExecutionResult.BlockID).Identity(receipt.ExecutorID)
+	if err != nil {
+		if protocol.IsIdentityNotFound(err) {
+			return false, engine.NewInvalidInputErrorf("could not get executor identity: %w", err)
+		}
+		// unknown exception
+		return false, fmt.Errorf("could not get executor identity: %w", err)
+	}
+
+	// for through each of the receipts to check for possible match of spocks
+	// all the spocks in a receipt will have to match in order to be counted as
+	// matched
+MatchingReceiptsLoop:
+	for _, u := range unmatchedReceipts {
+		// we check if the receipt chunks length matches, if they dont match we return
+		// false by default
+		if u.ExecutionResult.Chunks.Len() != receipt.ExecutionResult.Chunks.Len() {
+			return false, nil
+		}
+
+		// get receipt identity to get public key
+		uIdentity, err := v.protocolState.AtBlockID(u.ExecutionResult.BlockID).Identity(u.ExecutorID)
+		if err != nil {
+			if protocol.IsIdentityNotFound(err) {
+				return false, engine.NewInvalidInputErrorf("could not get executor identity: %w", err)
+			}
+			// unknown exception
+			return false, fmt.Errorf("could not get executor identity: %w", err)
+		}
+
+		// attempt to match every spock in the receipt with the candidate receipt
+		// if not verified then skip receipt
+		for _, chunk := range u.ExecutionResult.Chunks {
+			// check if spocks match
+			verified, err := crypto.SPOCKVerify(identity.StakingPubKey, receipt.Spocks[chunk.Index], uIdentity.StakingPubKey, u.Spocks[chunk.Index])
+			if err != nil {
+				return false, fmt.Errorf("could not verify spocks: %w", err)
+			}
+			if !verified {
+				continue MatchingReceiptsLoop
```

Comment:
Maybe better to decouple this into another method instead of having a loop label. We can have a tiny unexported method called for example `matchChunks` that given two receipts matches their chunks' SPoCKs.